test/PandoraInterface.cxx

```diff
@@ -257,40 +276,56 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
 
         int hitCounter(0);
 
+        // Create MCParticles from Geant4 trajectories
+        const MCParticleEnergyMap MCEnergyMap = CreateMCParticles(*pEDepSimEvent, pPrimaryPandora, parameters);
+
+        // Loop over (EDep) hits, which are stored in the hit segment detectors
         for (TG4HitSegmentDetectors::iterator detector = pEDepSimEvent->SegmentDetectors.begin();
              detector != pEDepSimEvent->SegmentDetectors.end(); ++detector)
         {
             std::cout << "Show hits for " << detector->first << " (" << detector->second.size() << " hits)" << std::endl;
+            std::cout << "                                 " << std::endl;
+
+            LArVoxelList voxelList;
 
-            for (TG4HitSegmentContainer::iterator g4Hit = detector->second.begin(); g4Hit != detector->second.end(); ++g4Hit)
+            // Loop over hit segments and create voxels from them
+            for (TG4HitSegment &g4Hit : detector->second)
             {
-                // const double length = g4Hit->GetTrackLength();
-                // const int contrib = g4Hit->Contrib.front();
-                // const std::string particle =
-                // pEDepSimEvent->Trajectories[contrib].GetName();
-
-                const double energy = g4Hit->GetEnergyDeposit();
-                const TLorentzVector &hitStart = g4Hit->GetStart() * mm2cm;
-                const TLorentzVector &hitStop = g4Hit->GetStop() * mm2cm;
-                const CartesianVector start(hitStart.X(), hitStart.Y(), hitStart.Z());
-                const CartesianVector stop(hitStop.X(), hitStop.Y(), hitStop.Z());
-                const CartesianVector centre((start + stop) * 0.5f);
+                LArVoxelList currentVoxelList = MakeVoxels(g4Hit, grid, parameters);
+
+                for (LArVoxel &voxel : currentVoxelList)
+                    voxelList.emplace_back(voxel);
+            }
+
+            std::cout << "Produced " << voxelList.size() << " voxels from " << detector->second.size() << " hit segments." << std::endl;
+
+            // Merge voxels with the same IDs
+            LArVoxelList mergedVoxels = MergeSameVoxels(voxelList);
+
+            std::cout << "Produced " << mergedVoxels.size() << " merged voxels from " << voxelList.size() << " voxels." << std::endl;
+
+            // Loop over the voxels and make them into caloHits
+            for (int i = 0; i < mergedVoxels.size(); i++)
+            {
+                const LArVoxel voxel = mergedVoxels[i];
```

Comment:
Looks to me like you can dispense with the loop counter for a range-based for loop of the form
for (const LArVoxel : mergeVoxels)